core/error.go

```diff
@@ -0,0 +1,29 @@
+package core
+
+import "github.com/vektah/gqlparser/v2/ast"
+
+type Error struct {
+	Query *Query `json:"-"`
+
+	Message string `field:"true" doc:"A description of the error."`
+
+	// Stack    string `field:"true" doc:"The stack trace of the error."`
+	// Source   *File  `field:"true" doc:"The source file where the error occurred."`
+	// Line     int    `field:"true" doc:"The line in the source file where the error occurred."`
+	// Original *Error `field:"true" doc:"The original error that caused this error."`
```

Comment:
Yeah this is probably the most interesting design point to dissect. It seems like your design is more geared towards source locations within modules, whereas I went for `*File` with things like linters in mind, which might run against an arbitrary source `*Directory`.

I went for `*File` since it felt the most powerful; any system interpreting the error can just read the file contents and display the errored line. For example, the TUI could show you the relevant code region.

But it might be better to have a context `*Directory` and a relative path beneath that directory.

```go
SourceDir *Directory
SourcePath string
```

That way you are able to see the full path, which is often interesting. And if there's no applicable `Directory` (go stdlib) maybe it can be left blank. (Though ideally even in the Go stdlib case we'd have a `Directory` somehow.)

Also, as a north star when thinking about this, it'd be nice to support things like [GitHub actions annotations](https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/workflow-commands-for-github-actions#setting-an-error-message) which require a path relative from repo root, which is only possible with a path - so that's another point against `File` I suppose.

> It would be cool if `Stack` was a `[]SourceMap`, then we can have a consistent stack trace between different languages, and can even combine/interleave them as needed 🎉

💯

Reply:
> But it might be better to have a context `*Directory` and a relative path beneath that directory.

That feels better to me, seems like it gets the best of both worlds? Not 100% how we'll get a `SourceDir` into something that's a `TypeDef` though. Maybe we could have like `SourceMap` know the `Module`/`ModuleSource` it was declared from? Then we could have a method to grab the `File`. Though now there's a circular dependency (is this an issue?)

I think we might need *something* like this anyways for supporting stack traces - the current module "name" I've implemented actually isn't good enough, since it's "local" to the module that consumes it - so stack traces that pass through multiple modules will not be super great.

I guess I'm suggesting altering #8498 from `Module string` to `Module *Module` - then we can have a `File` method easily.

---

Is the plan to do this in this PR, or to leave this commented for now and come back to it later?

Reply:
The plan is to leave it commented for now - the scope here is already way too broad. Some of my plans around the `Error` type are also starting to change with side discussions (e.g. error hierarchies might be redundant with returning a magic `Report` type derived from OTel spans instead), but the minimal implementation here is still good, I think.